Development.md

```diff
@@ -0,0 +1,250 @@
+...
+
+$ echo $?
+0
+```
+
+GitHub and Travis CI are configured to require that tests are passing before PR can be merged.
+
+The most complex tests, such as `TestProcessesListWithDetached` setup a stub AWS and Docker httptest web servers to simulate various request and response cycles. This can be challenging to write but represent a very powerful way to verify Convox behavior.
+
+## API Changes
+
+A common thing to do is to fix a bug or make an enhancement to the Rack APIs. For example, maybe the `POST /apps/{app}/builds` endpoint would be more helpful if it accepted a GitLab URL to clone and build from, so you'd like to add this.
+
+The `convox/rack/api` package has a few key concepts:
+
+* Swagger Manifest (rack/api/manifest.yml). Defines all API endpoints and responses.
+* Golang Client (rack/client). Bindings that talk to the HTTP API and returns Golang structs, slices and errors.
+* CLI (rack/cmd/convox). High level tool that lets developers issue commands like `convox/deploy`.
+* Routes (rack/api/controllers/routes). A `gorilla/mux` configuration of request URL patterns, HTTP verbs, and handler functions.
+* Controllers (rack/api/controllers). HTTP handlers for every route.
+* Models (rack/api/models). Key primatives like "app", "service", "build", and "release" and corresponding logic to control AWS and Docker.
+
+It is common for API changes to require corresponding changes across a model, controller, swagger manifest, client and CLI.
+
+When `convox start` successfully starts the Rack web, monitor, and registry processes locally, changes to the Golang source are detected, and the web process is rebuilt and restarted.
+
+## Systems Changes
+
+Many API calls need to execute changes across subsystems. For example:
+
+* `convox build` needs to create a new Docker container for the build and collect its output and return code
+* `convox release promote` needs to perform a CloudFormation stack update
+
+Systems engineering best practices are encouraged:
+
+* Robust error handling
+* Logging that makes a developers life easier
+* Logging that can be turned into operational metrics (e.g. `count#push.retry=1)`
+* Code strategies that make it easy to simulate subsystem requests/responses in a test environment
+
+## Infrastructure Changes
+
+Racks apps and services are created, updated and destroyed via automated means. This is a DevOps best practice that minimizes human errors and accidents that cause downtime. On AWS this is accomplished with CloudFormation. Some examples of changes:
+
+* Rack should have a new option to provision and use private subnets
+* An app load balancer should have a new option to configure Proxy Protocol
+* `convox services create elasticsearch` should provision an ElasticSearch cluster
+
+Some general notes when making changes to the infrastructure templates:
+
+* Run `make -C api templates` to compile the templates and restart the webserver. The `templates.go` file updates should be checked in.
```

Comment:
We should add a test / CI check / something that fails if templates.go is not current.